Fix/registration

RoyWilliams opened the pull request:
Changed the registration and change-password workflow to send an activation code by email to the user, rather than a link. The problem is that the link gets twisted up by the user's mail client and often doesn't work as planned.

RoyWilliams committed:
update

RoyWilliams committed:
update

RoyWilliams committed:
updte

gpfrancis commented:
I tested and it didn't work for me? I was able to login once, but not subsequently and looking at the database it looks like `is_active` did not get set.
```
MariaDB [ztf]> select * from auth_user where username='gftest';
+----+-------------------------------------------------------------------------------------------+------------+--------------+----------+------------+-----------+-----------------------------+----------+-----------+----------------------------+
| id | password                                                                                  | last_login | is_superuser | username | first_name | last_name | email                       | is_staff | is_active | date_joined                |
+----+-------------------------------------------------------------------------------------------+------------+--------------+----------+------------+-----------+-----------------------------+----------+-----------+----------------------------+
| 58 | pbkdf2_sha256$1000000$iYLznjyiDM7nGXpEFbsNfv$+tml6NrmTjo+9LpHPgaHDBI8raL4ajEE6fIqjh3ABCc= | NULL       |            0 | gftest   | Gareth     | Test      | pthime+lasairtest@gmail.com |        0 |         0 | 2025-06-30 10:21:49.087246 |
+----+-------------------------------------------------------------------------------------------+------------+--------------+----------+------------+-----------+-----------------------------+----------+-----------+----------------------------+
```

gpfrancis commented:
I don't think it's the cause of the above error, but I notice that you're using the builtin `hash()` to generate the hash. I don't think that's actually right because the output of that function isn't necessarily stable. Probably you should just use `hashlib.sha1`.

RoyWilliams committed:
new hashlib, import Token

RoyWilliams commented:
I changed the `hash` to `hashlib`. I added back the `Token` import for the profile. But I don't understand why the `gftest` user isn't active, it should be set [here](https://github.com/lsst-uk/lasair-lsst/blob/fix/registration/webserver/users/views.py#L108). I have deleted the user -- Gareth. can you try to register again?

gpfrancis commented:
OK, I've done some experimentation. It works if you do everything right, but there's a problem if you misstype the code in that you don't get the opportunity to correct it. However, you can still reset your password which then sends you an email containing a six digit code that is called an activation code in the email which does not actually activate the account (because it's really a validation code).

Basically it works if you do everything right, but we need to test all the ways that a user can possibly do it wrong and make sure those behave well too. And if we're asking users to type a number in instead of clicking on a link then they will definitely  type it wrong.

RoyWilliams commented:
Thanks Gareth. I fixed the "wrong code" error, now you just try again.
Please can you try other non-standard/ idiot workflows, and see where they go wrong?

gpfrancis commented:
I think there's a general problem with this flow in that if it gets interrupted for whatever reason - the email is too slow, user navigates away from the page, whatever - there doesn't appear to be any way to recover? You can't get back to the activation page, nor can you generate a new one.

Can we detect an attempt to log in with an inactive account and go back to reactivation from there?

gpfrancis commented:
Of course ideally we'd store the hashcode in the database rather than using a hidden form field, then we could just return to the page without having to send a new code. Also it would be more secure (although this is hardly a high security application).

RoyWilliams commented:
Surely the only actual problem is the user walking away in the middle of registration, leaving an inactive account? In that case, you try to login and it says "Please enter a correct username and password. " Let me think about this tomorrow.

RoyWilliams committed:
recovery for inactive users

RoyWilliams commented:
Another error fixed. If they register but walk away before doing the 6-digit code, then there is an "inactive user". They can't log in, but it tells them they are inactive, and says to use the Lost Password dialogue. If they complete that, with the 6-digit code, then their account becomes active.

gpfrancis commented:
Great! I think that fixes all the problems I identified.

Do note that what you've done isn't at all secure - I can send whatever activation code and hash I like and so long as they match they will be accepted - but for this application that doesn't really matter. If you want a low effort fix then you could add a salt to the hash which is only stored on the server side (maybe generate randomly at deploy time).

RoyWilliams commented:
OK how about the salt for the hash is kept in the vault with our other secrets. Can you arrange that?

gpfrancis commented:
No, I just tried it - if I replace the code I'm given with 123456 and the hash with 7c4a8d09ca3762af61e59520943dc26494f8941b then that works just as well as the ones I get sent. To be clear, I don't think that it matters in this application - doing that is more effort than just using a disposable email or something.

Just remember not to use this method in a context where we're trying to protect ourselves from a malicious user rather than a fumble-fingered user from themselves.

RoyWilliams committed:
added HASH_SALT

RoyWilliams commented:
I added `HASH_SALT` in the settings.py.
It should be set in deployment from a vault variable  called`hash_salt`
